File: src/test/java/org/artembogomolova/demo/webapp/test/domain/dao/repo/AbstractDaoTest.java
package org.artembogomolova.demo.webapp.test.domain.dao.repo;

import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import javax.validation.ConstraintViolationException;
import lombok.extern.slf4j.Slf4j;
import org.artembogomolova.demo.webapp.config.CustomHibernateValidatorConfiguration;
import org.artembogomolova.demo.webapp.config.ValidationConfig;
import org.artembogomolova.demo.webapp.domain.core.IdentifiedEntity;
import org.artembogomolova.demo.webapp.test.domain.AbstractDatabaseTest;
import org.artembogomolova.demo.webapp.validation.UniqueMultiColumnConstraint;
import org.artembogomolova.demo.webapp.validation.UniqueMultiColumnConstraint.UniqueMultiColumnConstraintColumns;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Import;
import org.springframework.data.repository.CrudRepository;
import org.springframework.test.context.TestExecutionListeners;
import org.springframework.test.context.support.DependencyInjectionTestExecutionListener;
import org.springframework.transaction.annotation.Transactional;

@TestExecutionListeners(value = {DependencyInjectionTestExecutionListener.class})
@ComponentScan("org.artembogomolova.demo.webapp.test.domain.dao.repo")
@Import({CustomHibernateValidatorConfiguration.class, ValidationConfig.class})
@Slf4j
public abstract class AbstractDaoTest<T extends IdentifiedEntity> extends AbstractDatabaseTest {

  protected static final Comparator ID_COMPARATOR = Comparator.comparing(IdentifiedEntity::getId);
  @Autowired
  protected EntityModifier entityModifier;
  private Class<T> clazz;

  protected AbstractDaoTest(Class<T> clazz) {
    this.clazz = clazz;
  }

  @Test
  @DisplayName("Testing for CRUD operation for entity.")
  @Transactional
  void crudCheck() {
    CrudRepository repository = getCrudRepository();
    try {
      log.info("build basic entities");
      List<T> entities = generateEntities();
      Assertions.assertFalse(entities == null || entities.isEmpty(), "entities list can't empty!");
      List<T> receivedCollection = entityModifier.transactionalSave(repository,
          entities,
          ID_COMPARATOR,
          (x, y) -> this.handleExceptions((Exception) x, (List<T>) y));
      log.info("basic entities saved in test db");
      entityModifier.transactionalUpdate(repository,
          receivedCollection,
          x -> {
            List<T> result = this.updateEntities((List<T>) x);
            log.info("basic entities modified");
            return result;
          },
          ID_COMPARATOR,
          (x, y) -> this.handleExceptions((Exception) x, (List<T>) y));
      log.info("basic entities updated in test db");
      repository.deleteAll(receivedCollection);
      Assertions.assertEquals(0, repository.count(), "exists entities not deleted from repository " + repository.findAll());
      validateAnotherRepositoryEmpty();
      log.info("passed!");
    } finally {
      if (repository != null) {
        repository.deleteAll();
      }
    }
  }

  protected boolean handleExceptions(Exception e, List<T> entities) {
    /*Exceptions are wrong.*/
    return false;
  }

  protected void validateAnotherRepositoryEmpty() {
    //no repositories validate needed
  }

  protected abstract List<T> updateEntities(List<T> savedCollection);

  protected abstract CrudRepository<T, Long> getCrudRepository();

  protected abstract List<T> generateEntities();

  @Test
  @Transactional
  @DisplayName("Testing for constraint violation exception for duplicate entity creating by unique constraint.")
  void duplicateDeniedTest() {
    UniqueMultiColumnConstraint uniqueMultiColumnConstraint = clazz.getAnnotation(UniqueMultiColumnConstraint.class);
    if (uniqueMultiColumnConstraint == null) {
      log.info("entity has no multi column constraint. passed!");
      return;
    }
    CrudRepository<T, Long> repository = getCrudRepository();
    try {
      Arrays.stream(uniqueMultiColumnConstraint.constraints()).forEach(
          uniqueMultiColumnConstraintColumns -> duplicateDeniedTestForConstraint(repository, uniqueMultiColumnConstraintColumns)
      );
      validateAnotherRepositoryEmpty();
      log.info("passed!");
    } finally {
      if (repository != null) {
        repository.deleteAll();
      }
    }
  }

  private void duplicateDeniedTestForConstraint(CrudRepository<T, Long> repository,
      UniqueMultiColumnConstraintColumns uniqueMultiColumnConstraintColumns) {
    Map<String, Object> commonValues = buildCommonFieldValues(uniqueMultiColumnConstraintColumns);
    if (commonValues == null) {
      commonValues = Collections.emptyMap();
    }
    prepareConditions(repository, uniqueMultiColumnConstraintColumns, commonValues);
    doDuplicateDeniedTestForConstraint(repository, uniqueMultiColumnConstraintColumns, commonValues);
  }

  protected abstract Map<String, Object> buildCommonFieldValues(
      UniqueMultiColumnConstraintColumns uniqueMultiColumnConstraintColumns);

  private void doDuplicateDeniedTestForConstraint(CrudRepository<T, Long> repository, UniqueMultiColumnConstraintColumns uniqueMultiColumnConstraintColumns,
      Map<String, Object> commonValues) {
    String uniqueMultiColumnConstraintColumnsName = uniqueMultiColumnConstraintColumns.name();
    /*second time to duplicate check*/
    try {
      tryToPersistEntityDuplicate(repository, uniqueMultiColumnConstraintColumns, commonValues);
    } catch (RuntimeException e) {
      if (!(e.getCause() instanceof ConstraintViolationException)) {
        throw e.getCause() != null ? new RuntimeException(e.getCause()) : e;
      }
      log.info("check for constraint '{}' passed!", uniqueMultiColumnConstraintColumnsName);
      return;
    }
    throw new RuntimeException("not implemented check for '" + uniqueMultiColumnConstraintColumnsName + "'");
  }

  protected void prepareConditions(CrudRepository<T, Long> repository, UniqueMultiColumnConstraintColumns uniqueMultiColumnConstraintColumns,
      Map<String, Object> commonValues) {
    tryToPersistPreparedEntity(repository, uniqueMultiColumnConstraintColumns, commonValues);
  }

  private void tryToPersistPreparedEntity(CrudRepository<T, Long> repository, UniqueMultiColumnConstraintColumns uniqueMultiColumnConstraintColumns,
      Map<String, Object> commonValues) {
    T entity = doPrepareDeniedTestEntity(uniqueMultiColumnConstraintColumns, commonValues);
    log.info("prepared entity {}", entity.toString());
    entityModifier.getSavedEntityCollection(repository,
        Collections.singletonList(entity),
        (x, y) -> this.handleExceptions((Exception) x, (List<T>) y));
  }

  protected abstract T doPrepareDeniedTestEntity(UniqueMultiColumnConstraintColumns uniqueMultiColumnConstraintColumns,
      Map<String, Object> commonValues);

  private void tryToPersistEntityDuplicate(CrudRepository repository,
      UniqueMultiColumnConstraintColumns uniqueMultiColumnConstraintColumns,
      Map<String, Object> commonValues) {
    T entity = doDuplicateDeniedTestEntity(uniqueMultiColumnConstraintColumns, commonValues);
    log.info("entity with duplicated column values {}", entity.toString());
    entityModifier.getSavedEntityCollection(repository,
        Collections.singletonList(entity),
        (x, y) -> this.handleExceptions((Exception) x, (List<T>) y));
  }

  protected abstract T doDuplicateDeniedTestEntity(UniqueMultiColumnConstraintColumns columns, Map<String, Object> commonValues);

}

File: app/src/test/kotlin/org/artembogomolova/demo/webapp/test/db/dao/AbstractDaoTest.kt
package org.artembogomolova.demo.webapp.test.db.dao

import java.util.*
import java.util.function.Supplier
import javax.validation.ConstraintViolation
import javax.validation.ConstraintViolationException
import javax.validation.Validator
import org.artembogomolova.demo.webapp.main.config.ValidationConfig
import org.artembogomolova.demo.webapp.main.config.db.CustomHibernateValidatorConfiguration
import org.artembogomolova.demo.webapp.main.domain.core.IdentifiedEntity
import org.artembogomolova.demo.webapp.main.validation.UniqueMultiColumn
import org.artembogomolova.demo.webapp.test.db.AbstractDatabaseTest
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.ComponentScan
import org.springframework.context.annotation.Import
import org.springframework.data.repository.CrudRepository
import org.springframework.transaction.annotation.Transactional

@ComponentScan("org.artembogomolova.demo.webapp.test.db.dao")
@Import(CustomHibernateValidatorConfiguration::class, ValidationConfig::class)
abstract class AbstractDaoTest<T : IdentifiedEntity<T>> protected constructor(
    clazz: Class<T>,
    private val constructor: Supplier<T>
) : AbstractDatabaseTest<T>(clazz, TEST_CLASS_NAME_SUFFIX, TEST_CLASS_DISPLAY_NAME_PREFIX) {
    @Autowired
    protected lateinit var entityModifier: EntityModifier<T>

    @Autowired
    private lateinit var validator: Validator

    @Test
    @DisplayName("Test for CRUD operation for entity.")
    @Transactional
    fun crudCheck() {
        val repository: CrudRepository<T, Long> = crudRepository()
        try {
            log.info("build basic entities")
            val entities = generateEntities()
            Assertions.assertFalse(entities.isEmpty(), "entities list can't empty!")
            val receivedCollection: List<T> = entityModifier.transactionalSave(
                repository,
                entities,
                ID_COMPARATOR
            ) { exception, list -> handleExceptions(exception, list) }
            log.info("basic entities saved in test db")
            entityModifier.transactionalUpdate(
                repository,
                receivedCollection as MutableList<T>,
                { x ->
                    val result: List<T> = updateEntities(x)
                    log.info("basic entities modified")
                    result
                },
                ID_COMPARATOR
            ) { exception, list -> handleExceptions(exception, list) }
            log.info("basic entities updated in test db")
            repository.deleteAll(receivedCollection)
            Assertions.assertEquals(EMPTY_REPOSITORY_COUNT, repository.count(), "exists entities not deleted from repository " + repository.findAll())
            validateAnotherRepositoryEmpty()
            log.info("passed!")
        } finally {
            cleanRepositories()
        }
    }

    /*Exceptions are wrong.*/
    protected fun handleExceptions(e: Exception?, entities: List<T>?): Boolean = false

    protected open fun validateAnotherRepositoryEmpty() {
        //no repositories validate needed
    }

    protected abstract fun updateEntities(savedCollection: List<T>): List<T>
    protected abstract fun crudRepository(): CrudRepository<T, Long>
    protected abstract fun generateEntities(): List<T>

    @Test
    @Transactional
    @DisplayName("Test for constraint violation exception for duplicate entity creating by unique constraint.")
    fun duplicateDeniedTest() {
        val uniqueMultiColumn: UniqueMultiColumn? = testingClass.getAnnotation(UniqueMultiColumn::class.java)
        if (uniqueMultiColumn == null) {
            log.info("entity has no multi column constraint. passed!")
            return
        }
        val repository: CrudRepository<T, Long> = crudRepository()
        try {
            Arrays.stream(uniqueMultiColumn.constraints)
                .forEach { uniqueMultiColumnConstraintColumns -> duplicateDeniedTestForConstraint(repository, uniqueMultiColumnConstraintColumns) }
            log.info("passed!")
        } finally {
            repository.deleteAll()
        }
    }

    private fun duplicateDeniedTestForConstraint(
        repository: CrudRepository<T, Long>,
        uniqueMultiColumnConstraint: UniqueMultiColumn.UniqueMultiColumnConstraint
    ) {
        var commonValues = buildCommonFieldValues(uniqueMultiColumnConstraint)
        if (commonValues == null) {
            commonValues = emptyMap<String, Any>()
        }
        prepareConditions(repository, uniqueMultiColumnConstraint, commonValues)
        doDuplicateDeniedTestForConstraint(repository, uniqueMultiColumnConstraint, commonValues)
        /*after each constraint test execution clean all used repositories!*/cleanRepositories()
    }

    protected fun cleanRepositories() {
        crudRepository().deleteAll()
    }

    protected abstract fun buildCommonFieldValues(
        uniqueMultiColumnConstraint: UniqueMultiColumn.UniqueMultiColumnConstraint
    ): Map<String, Any?>?

    private fun doDuplicateDeniedTestForConstraint(
        repository: CrudRepository<T, Long>, uniqueMultiColumnConstraint: UniqueMultiColumn.UniqueMultiColumnConstraint,
        commonValues: Map<String, Any?>
    ) {
        val uniqueMultiColumnConstraintColumnsName: String = uniqueMultiColumnConstraint.name
        /*second time to duplicate check*/try {
            tryToPersistEntityDuplicate(repository, uniqueMultiColumnConstraint, commonValues)
        } catch (e: RuntimeException) {
            if (e.cause !is ConstraintViolationException) {
                throw if (e.cause != null) RuntimeException(e.cause) else e
            }
            log.info("check for constraint '{}' passed!", uniqueMultiColumnConstraintColumnsName)
            return
        }
        throw RuntimeException("not implemented check for '$uniqueMultiColumnConstraintColumnsName'")
    }

    protected fun prepareConditions(
        repository: CrudRepository<T, Long>, uniqueMultiColumnConstraint: UniqueMultiColumn.UniqueMultiColumnConstraint,
        commonValues: Map<String, Any?>
    ) {
        tryToPersistPreparedEntity(repository, uniqueMultiColumnConstraint, commonValues)
    }

    private fun tryToPersistPreparedEntity(
        repository: CrudRepository<T, Long>, uniqueMultiColumnConstraint: UniqueMultiColumn.UniqueMultiColumnConstraint,
        commonValues: Map<String, Any?>
    ) {
        val entity = doPrepareDeniedTestEntity(uniqueMultiColumnConstraint, commonValues)
        log.info("prepared entity {}", entity.toString())
        entityModifier.getSavedEntityCollection(
            repository, listOf(entity)
        ) { exception, list -> handleExceptions(exception, list) }
    }

    protected abstract fun doPrepareDeniedTestEntity(
        uniqueMultiColumnConstraint: UniqueMultiColumn.UniqueMultiColumnConstraint,
        commonValues: Map<String, Any?>
    ): T

    private fun tryToPersistEntityDuplicate(
        repository: CrudRepository<T, Long>,
        uniqueMultiColumnConstraint: UniqueMultiColumn.UniqueMultiColumnConstraint,
        commonValues: Map<String, Any?>
    ) {
        val entity = doDuplicateDeniedTestEntity(uniqueMultiColumnConstraint, commonValues)
        log.info("entity with duplicated column values {}", entity.toString())
        entityModifier.getSavedEntityCollection(
            repository, listOf(entity)
        ) { x, y -> handleExceptions(x, y) }
    }

    protected abstract fun doDuplicateDeniedTestEntity(columns: UniqueMultiColumn.UniqueMultiColumnConstraint, commonValues: Map<String, Any?>): T

    @Test
    @DisplayName("Test for validation without another violations(not 'UniqueMultiColumnConstraint' annotations). Entity with all correct fields.")
    fun validatorTest() {
        val violations: Set<ConstraintViolation<T>> = validator.validate(buildEntityWithoutViolationEntity())
        Assertions.assertTrue(violations.isEmpty(), "entity has following violations: $violations")
    }

    protected open fun buildEntityWithoutViolationEntity(): T = constructor.get()

    @Test
    @DisplayName(
        "Test for validation with another violations(not 'UniqueMultiColumnConstraint' annotations). "
                + "Entity with one or more incorrect fields. Basically entity no-arg constructor"
    )
    fun validatorWithViolationTest() {
        val violations: Set<ConstraintViolation<T>> = validator.validate(buildWithViolationEntity())
        log.info("entity has following violations: $violations")
        Assertions.assertFalse(violations.isEmpty(), "entity has no violations!")
        /* val exceptedViolations = buildExceptedViolations()
         Assertions.assertEquals(exceptedViolations, violations, "excepted violations $exceptedViolations, actual violations: $violations\"")*/
    }

//    protected open fun buildExceptedViolations(): Set<ConstraintViolation<T>> = setOf()

    protected fun buildWithViolationEntity(): T = constructor.get()

    companion object {
        private const val TEST_CLASS_NAME_SUFFIX = "DaoTest"
        private const val TEST_CLASS_DISPLAY_NAME_PREFIX = "Dao test: "
        private val ID_COMPARATOR = Comparator.comparing<IdentifiedEntity<*>, Long?>(IdentifiedEntity<*>::id)
        internal const val EMPTY_REPOSITORY_COUNT = 0L
    }
}